src/common/storage/backends/s3/Client.js

```diff
@@ -0,0 +1,199 @@
+        await this.stageConfigAndCreateNewS3Client(dataInfo);
+        const {bucketName, filename} = dataInfo.data;
+        const s3Client = await this.getS3Client();
+        const data = await s3Client.getObject({
+            Bucket: bucketName,
+            Key: filename
+        });
+        await this.restoreS3Client();
+        return data.Body;
+    };
+
+    S3Storage.prototype.putFile = async function (filename, content) {
+        await this.createBucketIfNeeded();
+        this.logger.debug(`Created bucket ${this.bucketName}`);
+        const params = {
+            Body: require.isBrowser ? new Blob([content]) : content,
+            Bucket: this.bucketName,
+            Key: filename,
+        };
+        const s3Client = await this.getS3Client();
+        await s3Client.putObject(params);
+        const metadata = await this._stat(filename, this.bucketName);
+        metadata.filename = filename;
+        metadata.size = metadata.ContentLength;
+        metadata.bucketName = this.bucketName;
+        metadata.endpoint = this.config.endpoint;
+        this.logger.debug(`Successfully uploaded file ${filename} to the S3 server.`);
+        return this.createDataInfo(metadata);
+    };
+
+    S3Storage.prototype._stat = async function (path, bucketName) {
+        const params = {
+            Bucket: bucketName,
+            Key: path
+        };
+        return (await this.getS3Client()).headObject(params);
+    };
+
+    S3Storage.prototype.deleteDir = async function (dirname) {
+        let dataInfo;
+        const s3Client = await this.getS3Client();
+        const {Contents} = await s3Client.listObjectsV2({
+            Bucket: this.bucketName,
+            MaxKeys: 1000,
+            Prefix: dirname
+        });
+
+        for (const file of Contents) {
+            dataInfo = {
+                data: {
+                    bucketName: this.bucketName,
+                    filename: file.Key,
+                    endpoint: this.endpoint
+                }
+            };
+            await this.deleteFile(dataInfo);
+        }
+        this.logger.debug(`Successfully deleted directory ${dirname} from the S3 server`);
+    };
+
+    S3Storage.prototype.deleteFile = async function (dataInfo) {
+        await this.stageConfigAndCreateNewS3Client(dataInfo);
+        const {bucketName, filename} = dataInfo.data;
+        const params = {
+            Bucket: bucketName,
+            Key: filename
+        };
+        await this.s3Client.deleteObject(params);
+        await this.restoreS3Client();
+    };
+
+    S3Storage.prototype.stageConfigAndCreateNewS3Client = async function(dataInfo) {
+        this.stagedConfig = JSON.parse(JSON.stringify(this.config));
+        const {endpoint} = dataInfo.data;
+        const {accessKeyId, secretAccessKey} = this.config;
+        this.config = this.createS3Config({endpoint, accessKeyId, secretAccessKey});
+        this.ready = await this.initialize();
```

Comment:
It might be more clear (and a bit cleaner) to make this a bit less stateful. In its current state, it could actually have some difficult to hunt down bugs due to this staging during async methods. (Suppose there is a default config, say config1, and two calls to `getFile` which each create two additional configs, config2 and config3. If you call `getFile` twice, the second call will overwrite the staged config and config1 will be lost.)

I would recommend updating `getS3Client` to accept an optional config. If none is provided, it can default to `this.config`. Then you can actually get rid of `this.s3Client` entirely and just instantiate a new client on each call.

Reply:
We can also get rid of all the staging and restoring logic that way, too.

Reply:
But wouldn't it mean that we will be creating `S3` Objects more than necessary? I agree that staging logic could be improved and kind of not necessary if we remove the class level reference to `s3Client`.

Reply:
`getS3Client` will now operate in the following Manner:
1. Wait till `aws-sdk` has been loaded.
2. If there's no config(use this.config), check if the current client is null, of it is null create a new `s3client` and assign it to this.s3Client. Return this.s3Client.
3. If `config` is provided create a new client based on the config and return the client object.